### DISCUSSION TITLE
Multi-tenant architecture based on Roda + Rodauth + Sequel

### DISCUSSION THREAD

**Discussion by okliv** [jeremyevans/roda#289] [General]
i have one "master" database with users for `rodauth` and all other separate user-specific databases

what i heaped up for now to implement it (it works, but i worry that this is ugly design and more elegant solution exists):

```ruby
# app.rb

AUTH_DB ||= Sequel.sqlite(database: 'auth_db.sqlite3')

class App < Roda
  #...
  plugin :rodauth do
    db AUTH_DB
  #...
  def current_user
      @current_user ||= Account[rodauth.account_from_session[:id]] if rodauth.logged_in?
  end
  #...
  before do
    if current_user
     # intuitively ugly part
      Model ||= Class.new(Sequel::Model(current_user.db)) do

```
```ruby
# account.rb

class Account < Sequel::Model(AUTH_DB)
    def db
      @db ||=  Sequel.sqlite("#{id}_db.sqlite3")
    end

```
and so, i can now pass any other user-specific data through models like this
```ruby
# user_something.rb

class UserSomething < App::Model
```
is this ok or can you advice a better solution/plugin (except single-tenancy)? thanks

**Comment by jeremyevans** [comment]
> i have one "master" database with users for `rodauth` and all other separate user-specific databases
> 
> what i heaped up for now to implement it (it works, but i worry that this is ugly design and more elegant solution exists):

I'm surprised that some parts here work.

> ```ruby
> # app.rb
> 
> AUTH_DB ||= Sequel.sqlite(database: 'auth_db.sqlite3')

The `||=` shouldn't be necessary unless you are reloading this file or are setting `AUTH_DB` in some way before requiring it.  Unless one of those things are true, using `=` should be fine.

> 
> class App < Roda
> #...
>   plugin :rodauth do
>     db AUTH_DB
> #...
>   def current_user
>       @current_user ||= Account[rodauth.account_from_session[:id]] if rodauth.logged_in?

This seems kind of strange.  Have you seen what query it generates?  Maybe it would be better to do:

  @current_user ||= Account.with_pk!(rodauth.session_value) if rodauth.logged_in?

>   end
> #...
>   before do
>     if current_user
>      # intuitively ugly part
>       Model ||= Class.new(Sequel::Model(current_user.db)) do

This seems like it wouldn't work.  There would only be one Model constant, using the database of the first user.  Using a constant for this is probably wrong.  At the very least, if you are using a constant, you'd want to hash with user ids and model class values.

> is this ok or can you advice a better solution/plugin (except single-tenancy)? thanks

It sounds like you may want to use Sequel's sharding support instead of something like this, assuming all databases have the same schema.

**Reply by okliv** [reply]
> > i have one "master" database with users for `rodauth` and all other separate user-specific databases
> > what i heaped up for now to implement it (it works, but i worry that this is ugly design and more elegant solution exists):
> 
> I'm surprised that some parts here work.
> 
> > ```ruby
> > # app.rb
> > 
> > AUTH_DB ||= Sequel.sqlite(database: 'auth_db.sqlite3')
> > ```
> 
> The `||=` shouldn't be necessary unless you are reloading this file or are setting `AUTH_DB` in some way before requiring it. Unless one of those things are true, using `=` should be fine.

awh. yeah, it left from some experiments

> 
> > class App < Roda
> > #...
> > plugin :rodauth do
> > db AUTH_DB
> > #...
> > def current_user
> > @current_user ||= Account[rodauth.account_from_session[:id]] if rodauth.logged_in?
> 
> This seems kind of strange. Have you seen what query it generates? Maybe it would be better to do:
> 
> @current_user ||= Account.with_pk!(rodauth.session_value) if rodauth.logged_in?

I didn't found any documentation of this, just dug it through debug console, thanks

> 
> > end
> > #...
> > before do
> > if current_user
> > # intuitively ugly part
> > Model ||= Class.new(Sequel::Model(current_user.db)) do
> 
> This seems like it wouldn't work. There would only be one Model constant, using the database of the first user. Using a constant for this is probably wrong. At the very least, if you are using a constant, you'd want to hash with user ids and model class values.

it is intranet application with small amount of users. At least two of them surely have different data while ~simultaneously using the app. Here `Model` constant recreates at runtime (Class.new) and looks like it can be not thread-safe... from the other hand If i hash user ids into it, than i will not be able to address it with `< App::Model`

> 
> > is this ok or can you advice a better solution/plugin (except single-tenancy)? thanks
> 
> It sounds like you may want to use Sequel's sharding support instead of something like this, assuming all databases have the same schema.

Will sharding plugin allow me to write different data to different shards and potentially allow me to modify structure of some (not all) user databases?

**Reply by jeremyevans** [reply]
> it is intranet application with small amount of users. At least two of them surely have different data while ~simultaneously using the app. Here `Model` constant recreates at runtime (Class.new) and looks like it can be not thread-safe... from the other hand If i hash user ids into it, than i will not be able to address it with `< App::Model`

You should probably use an approach like this:

```ruby
MODELS = {}
# ...
before do
  if current_user
    @model = MODELS[current_user.id] ||= begin
      c = Class.new(App::Model)
      c.dataset = current_user.db[:table_name]
      c
    end
  end
end
```

With a mutex if you want it to be thread-safe.

> Will sharding plugin allow me to write different data to different shards and potentially allow me to modify structure of some (not all) user databases?

No, sharding should only be used if the schema is the same on all databases, so it would not be appropriate for this use case.

**Reply by okliv** [reply]
> The `||=` shouldn't be necessary

i found why it is so: because if i use just `=` i have this annoying stdout message... (i use `zeitwerk`)

![image](https://user-images.githubusercontent.com/884634/200422511-4bfcf697-5839-4a47-b58b-7f9a16251012.png)

**Reply by jeremyevans** [reply]
That's because zeitwerk doesn't remove the constant.  If you move the constant under App (so it is App::AUTH_DB), that's an alternative way to fix it that should work with zeitwerk.

With your current setup, you aren't setting up a new Database object when you reload the file.  If you make a change to the Database object (adding an option for example), it wouldn't get picked up.

**Comment by okliv** [comment]
> This seems like it wouldn't work. There would only be one Model constant, using the database of the first user. Using a constant for this is probably wrong.

you were right. i had to mention that we have `zeitwerk`'s reloading feature enabled.

if i disable this feature (like we normally do in production environment) everything stops to work as you noticed.

i found that zeitwerk has it's own [thread safety fuse](https://github.com/fxn/zeitwerk#thread-safety) but i still worry if this actually an appropriate setup, when i reload models on each request...

If i assign model as an instance variable i have lot of cumbersomeness with associations and other Sequel::Model specific features that i need to rewrite, or [path plugin](https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/Path.html) stops to work which i used to define class-level paths...

In other words, if we forget about my current setup and if it is possible, how would you advice to design multi-tenant architecture based on Roda + Rodauth + multiple Sequel model classes with a lot of model level logic? Is it designed for this case (if i can stand a question in this way)?

Thanks a lot

**Reply by jeremyevans** [reply]
It's hard to answer this question.  I'm not sure there is a good way with the database design you have chosen.  In general, for multi-tenant, you have three decent ways:

1) Store data for multiple tenants in the same table, with a tenant_id or similar column for the tenant
2) Store data for each tenant in separate tables per tenant, with the same schema for the tables for all tenants
3) Use a separate application/process per tenant, with possibly different schemas per tenant

Option 1 is fairly simple, but you need to be careful to enforce your per-tenant filter for all queries.  Option 2 is also fairly simple to implement using Sequel's sharding support.  Option 3 requires more resources, but is probably simplest at the application level since you don't need to do per-tenant filtering or sharding.  The challenge with option 3 is orchestrating/managing the large number of processes.

You seem to want a combination of 2 and 3, but it is more work to get such an approach to work well.  The best you can probably do is a separate Sequel::Database per tenant, and separate Sequel::Model classes per tenant.

If I had to stick to your database design, I'd probably try to move to option 3.  As long as the revenue per tenant exceeds the costs of running the tenant's application, such an approach should be feasible.  However, if I had to design something similar from scratch, I'd probably go with option 1, unless one of the requirements was having all data separated per tenant, in which case I'd use option 2.
